### PR TITLE
feat(web): auto switch theme

### PR DESCRIPTION
**Pull request by martabal**
## Changes made in this PR

This PR adds the ability to automatically switch the web UI theme based on the browser preferences.

## Screenshots

https://github.com/immich-app/immich/assets/74269598/d27d50f2-1c2f-47bf-9d6e-c0aa4151065c

### CODE REVIEW

**Review by jrasm91 — COMMENTED** [review]
This still looks overly complicated to me, especially when it is essentially a pretty simple feature. Can we not do something like this?
1. Detect the theme on start-up and _initialize a store_ in the main layout.
2. List to the store and synchronize the document class in the main layout.
3. Use the store to determine if the theme switcher should be shown or not.
4. Keep the app.html duplicate stuff to prevent flashing

For me it is hard to follow all the events and reactions. I would appreciate if it was streamlined.

File: web/src/lib/utils/browser-utils.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { Theme } from '$lib/constants';
+import { colorTheme } from '$lib/stores/preferences.store';
+import { get } from 'svelte/store';
+
+export const handleToggleTheme = () => {
+  const theme = get(colorTheme);
+  theme.value = theme.value === Theme.DARK ? Theme.LIGHT : Theme.DARK;
+  colorTheme.set(theme);
+};
```
Comment (jrasm91) [inline web/src/lib/utils/browser-utils.ts:9]: This should probably just exist in the same file as the store.

**Review by jrasm91 — APPROVED** [review]
Looks better.

File: web/src/lib/stores/preferences.store.ts [hunk 2]
```diff
@@ -1,13 +1,38 @@
...
 
 // The 'color-theme' key is also used by app.html to prevent FOUC on page load.
-export const colorTheme = persisted<'dark' | 'light'>('color-theme', initialTheme, {
+export const colorTheme = persisted<ThemeSetting>('color-theme', initialTheme, {
   serializer: {
-    parse: (text) => (text === 'light' ? text : 'dark'),
-    stringify: (obj) => obj,
+    parse: (text: string): ThemeSetting => {
+      const parsedText: ThemeSetting = JSON.parse(text);
+      if (parsedText.system === undefined) {
```
[the remark this answers is not in the document]
Comment (martabal) [reply web/src/lib/stores/preferences.store.ts:26]: It feels wrong to only check the theme value and not the theme system, but unless you modify yourself the local storage variable, it shouldn't be a problem

[the remark this answers is not in the document]
Comment (jrasm91) [reply web/src/lib/stores/preferences.store.ts]: Then I would prefer to remove this.

[the remark this answers is not in the document]
Comment (martabal) [reply web/src/lib/stores/preferences.store.ts]: Right, I already removed it

File: web/src/lib/stores/preferences.store.ts [hunk 3]
```diff
@@ -1,13 +1,42 @@
...
 
 // The 'color-theme' key is also used by app.html to prevent FOUC on page load.
-export const colorTheme = persisted<'dark' | 'light'>('color-theme', initialTheme, {
+export const colorTheme = persisted<ThemeSetting>('color-theme', initialTheme, {
   serializer: {
-    parse: (text) => (text === 'light' ? text : 'dark'),
-    stringify: (obj) => obj,
+    parse: (text: string): ThemeSetting => {
+      const parsedText: ThemeSetting = JSON.parse(text);
+      if (Object.values(Theme).includes(parsedText.value as Theme)) {
```
Comment (danieldietzler) [inline web/src/lib/stores/preferences.store.ts:33]: Is that `as Setting` still necessary since `parsedText` is already of type `ThemeSetting`?

**Review by danieldietzler — APPROVED** [review]
LGTM!

File: web/src/lib/stores/preferences.store.ts (hunk 3 above)
Comment (martabal) [reply web/src/lib/stores/preferences.store.ts]: You're right, it's not needed. Thanks !

**Review by jrasm91 — APPROVED** [review]
